### PR TITLE
GH-103525: Improve exception message from `pathlib.PurePath()`

### PR DESCRIPTION
**Pull request by barneygale**
Check that arguments are strings before calling `os.path.join()`.

Also improve performance of `PurePath(PurePath(...))` while we're in the area: we now use the *unnormalized* string path of path arguments.

* Issue: gh-103525

### CODE REVIEW

File: Lib/pathlib.py [hunk 1]
```diff
@@ -300,18 +300,24 @@ def __reduce__(self):
...
+        for arg in args:
+            if isinstance(arg, PurePath):
+                path = arg._raw_path
+            else:
+                path = os.fspath(arg)
+                if not isinstance(path, str):
+                    raise TypeError(
+                        "argument should be a str or an os.PathLike "
+                        "object where __fspath__ returns a str, "
+                        f"not {type(path).__name__!r}")
```
Comment (terryjreedy) [inline Lib/pathlib.py]: Give improved message for any bad argument.  See edited comment on issue.
```suggestion
                try:
                    path = os.fspath(arg)
                except TypeError:
                    path = arg
                if not isinstance(path, str):
                    raise TypeError(
                        "argument should be a str or an os.PathLike "
                        "object where __fspath__ returns a str, "
                        f"not {type(path).__name__!r}")
```

Comment (barneygale) [reply Lib/pathlib.py]: I didn't see this solution. Very nice :)

Comment (akulakov) [reply Lib/pathlib.py]: I think it should be "an str" following this rule: https://www.btb.termiumplus.gc.ca/tpv2guides/guides/wrtps/index-eng.html?lang=eng&lettr=indx_catlog_a&page=9MN1DTkGtnPE.html#:~:text=When%20the%20indefinite%20article%20is,before%20the%20consonant%20sound%20n)

Even though `str` is not an acronym, it's pronounced as an acronym, i.e. "as tee arr". I don't know if there was a discussion about this but there's a couple places in the docs where "a str" is used:
library/io.rst:      Return a ``str`` containing the entire contents of the buffer.
library/importlib.resources.rst:    Read and return the contents of *resource* within *package* as a ``str``.

And one place were "an str" is used.

Comment (terryjreedy) [reply Lib/pathlib.py]: 'an str' looks completely wrong to me.  'str' is an abbreviation for 'string' and is neither an acronym or initialism, the subject of the reference.  I *read* it, at least silently, as 'string' or possibly 'stir', not as 'es tee are'.  I would change the occurrence of 'an str'.

**Comment by barneygale** [comment]
Hey @terryjreedy, do you think this is ready to land? Thank you!

**Comment by terryjreedy** [comment]
Assuming that the deletions of '._raw_path' in the dunders is correct (I skipped over these), yes.

**Pull request merged**